[Lee Yat Bun] iP

yatbun opened the pull request:
# Duke Pro MAX (TM)

> "The bestest Duke there is out there" – My grandfather's son's son

Duke Pro MAX (TM) is the next biggest thing in the personal organisation space. It's:
- text-based 📰
- easy to learn 🤓
- ~~FAST~~ *SUPER DUPER* FAST to learn 🚀

To get started today, all you need to do is:
1. Download the latest and greatest JAR file from the [releases page](https://github.com/yatbun/ip/releases),
2. Double-click it,
3. Add your tasks,
4. Let it manage the tasks for you!

And what's more, **ITS 100% FREE**!

Features include but are not limited to:
- [x] Managing Tasks
- [x] Managing Deadlines
- [ ] Reminders (_coming soon_)

---

## FAQ

**Q: How does Duke Pro MAX (TM)'s main function look like?**
A: 
```
public static void main(String[] args) {
    new Duke("data/duke.txt").run();
}
```

Commit:
Add Gradle support

yatbun committed:
Implemented skeletal version of Duke
- Greets, echos and exits upon command

yatbun committed:
Added the ability to store and display user's inputs

yatbun committed:
Added the ability to mark tasks as done.

yatbun committed:
Added support for ToDos, Events and Deadlines.

yatbun committed:
Implemented a test using the I/O redirection technique.

yatbun committed:
Implemented error handling.

yatbun committed:
Added support for deleting tasks.

yatbun committed:
Standardised ToDo implementation

yatbun committed:
Implement data saving and loading

yatbun committed:
Deadline class: Add LocalDate support

yatbun committed:
Merge from yatbun/branch-Level-7

Implement data saving and loading

yatbun committed:
Merge branch 'master' into branch-Level-8

yatbun committed:
Merge from yatbun/branch-Level-8

Deadline class: Add LocalDate support

yatbun committed:
Deadlines class: Fixed date conversion bug

yatbun committed:
Refactor code to be more OOP

yatbun committed:
Divide classes into packages

yatbun committed:
Add JUnit tests

yatbun committed:
Package the App as a JAR file

yatbun committed:
Add JavaDoc comments

yatbun committed:
Parser class: Tweaked code to comply to coding standard

yatbun committed:
Implement searching of tasks

yatbun committed:
Merge from yatbun/branch-A-JavaDoc

Add JavaDoc comments

yatbun committed:
Merge from yatbun/branch-A-CodingStandard

Parser class: Tweaked code to comply to coding standard

yatbun committed:
Merge branch 'master' into branch-Level-9

yatbun committed:
Merge from yatbun/branch-Level-9

Implement searching of tasks

yatbun committed:
Merge remote-tracking branch 'origin/add-gradle-support' into A-Gradle

yatbun committed:
Add Gradle to project

yatbun committed:
Fix code styling according to Checkstyle

yatbun committed:
Update checkstyle configuration

yatbun committed:
Merge from yatbun/A-Gradle

Merge A-Gradle branch into master

Zzhiwei commented on src/main/java/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,28 @@
+     * @param userInput The raw input from the user
+     */
+    public DoneCommand(String userInput) {
+        this.taskNumber = Integer.parseInt(
```
if there's no ambiguity, should we still use "this"?

Zzhiwei commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Runs the program until termination.
+     */
+    public void run() {
+        ui.showWelcomeMessage();
+
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String input = ui.getUserInput();
+                Command c = parser.parseCommand(input);
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (DukeException e) {
+                ui.showErrorMessage(e);
+            }
+        }
+
+        ui.showGoodbyeMessage();
```
I like the separation of code groups!

Zzhiwei commented on src/main/java/duke/commands/CreateTodoCommand.java:
```diff
@@ -0,0 +1,26 @@
+public class CreateTodoCommand extends Command {
+    private final String name;
+
+    /**
```
should you also give a brief summary of what the method does first in javadocs first line?

Zzhiwei commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+        this.searchString = userInput.substring(5);
+    }
+
+    @Override
```
shouldn't we also write javadocs for subclass methods even if overriden? I think there's a intellij function for this, you can search it up on google.

Zzhiwei commented on src/main/java/duke/data/TaskList.java:
```diff
@@ -0,0 +1,85 @@
+
+    /**
+     * Adds a Task to the list.
+     * @param t The task to be added
```
should we leave a space between @param and description?

Zzhiwei commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,76 @@
+ * Parsers user input.
+ */
+public class Parser {
+    private static final Pattern DONE_FORMAT =
+            Pattern.compile("^done\\s\\d+");
+
+    private static final Pattern DELETE_FORMAT =
+            Pattern.compile("^delete\\s\\d+");
+
+    private static final Pattern FIND_FORMAT =
+            Pattern.compile("^find\\s.+");
+
+    private static final Pattern TODOS_FORMAT =
+            Pattern.compile("^todo\\s.+");
+
+    private static final Pattern DEADLINES_FORMAT =
+            Pattern.compile("^deadline\\s.+\\s/by\\s.+");
+
+    private static final Pattern EVENTS_FORMAT =
+            Pattern.compile("^event\\s.+\\s/at\\s.+");
```
nicely done!

Zzhiwei reviewed (changes requested):
Really clean code!

yatbun committed:
Add a GUI to Duke

yatbun committed:
Complete add of GUI to Duke

yatbun committed:
Merge from yatbun/branch-Level-10

Add GUI to Duke

yatbun committed:
Build jar file with JavaFX dependencies

yatbun committed:
Add assertions to Commands

Commands take in user input and extracts information to create the
corresponding task.

Let's add assertions to places where user input are being split in case
the resultant length of the array is shorter than expected.

yatbun committed:
Improved quality of code

"Happy path" was not very prominent for some parts of parser.

We should add spacing around error cases to distiguish the happy path.

yatbun committed:
Merge from yatbun/branch-A-CodeQuality

Improved quality of code

yatbun committed:
Merge from yatbun/branch-A-Assertions

Add assertions to Commands

yatbun committed:
Add tagging of tasks

yatbun committed:
Merge from yatbun/branch-C-Tagging

Add tagging of tasks

yatbun committed:
Improve the GUI

yatbun committed:
Merge branch 'master' of https://github.com/yatbun/ip

yatbun committed:
Add a representative screenshot

yatbun committed:
Add a User Guide
